tools/pkg/constraints.toit

```diff
@@ -15,29 +15,60 @@
 
 import .semantic-version
 import .parsers.constraint-parser
+import .parsers.semantic-version-parser
 
 class Constraint:
-  constraints/List
+  simple-constraints/List
   source/string
 
-  constructor .source/string:
+  constructor --.simple-constraints --.source:
+
+  static parse source/string -> Constraint:
     parsed := (ConstraintParser source).constraints --consume-all
-    constraints = []
+    constraints := []
     parsed.do: | constraint/ConstraintParseResult |
       version := SemanticVersion.from-parse-result constraint.semantic-version
       if constraint.prefix == "^":
-        constraints.add (SimpleConstraint ">=" version)
-        constraints.add (SimpleConstraint "<" (SemanticVersion --major=version.major + 1))
+        // All versions compatible with the given version.
+        if version.major != 0:
+          constraints.add (SimpleConstraint ">=" version)
+          constraints.add (SimpleConstraint "<" (SemanticVersion --major=version.major + 1))
+        else:
+          // If the major version is 0, then the minor version may not be increased.
+          constraints.add (SimpleConstraint ">=" version)
+          constraints.add (SimpleConstraint "<" (SemanticVersion --major=version.major --minor=version.minor + 1))
       else if constraint.prefix == "~" or constraint.prefix == "~>":
         constraints.add (SimpleConstraint ">=" version)
         constraints.add (SimpleConstraint "<" (SemanticVersion --major=version.major --minor=version.minor + 1))
       else if constraint.prefix == "":
         constraints.add (SimpleConstraint "=" version)
       else:
         constraints.add (SimpleConstraint constraint.prefix version)
+    return Constraint --simple-constraints=constraints --source=source
+
+  static parse-range source/string -> Constraint:
+    parser := SemanticVersionParser source --allow-missing-minor
+    parse-result := parser.semantic-version --consume-all
+    // triple/TripleParseResult
+    // pre-releases/List
+    // build-numbers/List
+    triple := parse-result.triple.triple
```

Comment:
This looks kinda silly. Why not have major, minor, and patch on the TripleParseResult class?

Reply:
I'm not super happy with the structure either.
Will put it on my list of refactoring TODOs.